File: 8.py
class Solution:
    # @return an integer
    def atoi(self, str):
        buf = 0
        idx = 0
        neg = False
        INT_MAX = 2147483647
        INT_MIN = -2147483648
        INT_MAX_10 = INT_MAX / 10
        INT_MIN_10 = INT_MIN / 10
        # skip space
        str = str.strip()
        LL = len(str)
        if LL <= 0:
            return 0
        # get sign
        if str[idx] == '-': # "-123"
            neg = True
            idx += 1
        elif str[idx] == '+' : # '+123'
            idx += 1
        # handle numbers
        if neg:
            while idx < LL and '0' <= str[idx] <= '9':
                if buf < INT_MIN_10:
                    buf = INT_MIN
                    break
                buf *= 10
                tmp = -int(str[idx])
                if buf < INT_MIN - tmp:
                    buf = INT_MIN
                    break
                buf += tmp
                idx += 1
        else:
            while idx < LL and '0' <= str[idx] <= '9':
                if buf > INT_MAX_10:
                    buf = INT_MAX
                    break
                buf *= 10
                tmp = int(str[idx])
                if buf > INT_MAX - tmp:
                    buf = INT_MAX
                    break
                buf += tmp
                idx += 1
        return buf

File: 8.java
public class Solution {
    public int atoi(String str) {
        int buf = 0;
        int idx = 0, LL = str.length();
        boolean neg = false;
        int tmp = 0;
        int INT_MAX = Integer.MAX_VALUE;
        int INT_MIN = Integer.MIN_VALUE;
        int INT_MAX_10 = INT_MAX / 10;
        int INT_MIN_10 = INT_MIN / 10;
        // skip space
        for (; idx < LL && str.charAt(idx) == ' '; idx++);
        // get sign
        if (idx < LL && str.charAt(idx) == '-') { // "-123"
            neg = true;
            idx++;
        } else if (idx < LL && str.charAt(idx) == '+') { // '+123'
            idx++;
        }
        // handle numbers
        if (neg) {
            for (; idx < LL && Character.isDigit(str.charAt(idx)); idx++) {
                if (buf < INT_MIN_10) {
                    buf = INT_MIN;
                    break;
                }
                buf *= 10;
                tmp = -Character.digit(str.charAt(idx), 10);
                if (buf < INT_MIN - tmp) {
                    buf = INT_MIN;
                    break;
                }
                buf += tmp;
            }
        } else {
            for (; idx < LL && Character.isDigit(str.charAt(idx)); idx++) {
                if (buf > INT_MAX_10) {
                    buf = INT_MAX;
                    break;
                }
                buf *= 10;
                tmp = Character.digit(str.charAt(idx), 10);
                if (buf > INT_MAX - tmp) {
                    buf = INT_MAX;
                    break;
                }
                buf += tmp;
            }
        }
        return buf;
    }
}